src/audio/pipewire.c

```diff
@@ -0,0 +1,254 @@
+    return (AudioID *)state;
+}
+// configure a pipewire stream for the given speech dispatcher configuration and then prepair loop for playback
+static int pipewire_begin(AudioID *id, AudioTrack track)
+{
+    module_state *state = (module_state *)id;
+    const struct spa_pod *params[1];
+    uint8_t buffer[1024];
+    struct spa_pod_builder b = SPA_POD_BUILDER_INIT(buffer, sizeof(buffer));
+    int format;
+    state->channel_count = track.num_channels;
+    if (track.bits == 16)
+    {
+        switch (state->id.format)
+        {
+        case SPD_AUDIO_LE:
+            format = SPA_AUDIO_FORMAT_S16_LE;
+            break;
+        case SPD_AUDIO_BE:
+            format = SPA_AUDIO_FORMAT_S16_BE;
+            break;
+        default:
+            pw_log_error("invalid audio format specifier");
+            return -1;
+        }
+    }
+    else if (track.bits == 8)
+    {
+        format = SPA_AUDIO_FORMAT_S8;
+    }
+    else
+    {
+        pw_log_error("invalid audio format specifier");
+        return -1;
+    }
+
+    params[0] = spa_format_audio_raw_build(&b, SPA_PARAM_EnumFormat, &SPA_AUDIO_INFO_RAW_INIT(.format = format, .channels = track.num_channels, .rate = track.sample_rate));
+    pw_stream_connect(state->stream, PW_DIRECTION_OUTPUT, PW_ID_ANY, PW_STREAM_FLAG_AUTOCONNECT | PW_STREAM_FLAG_MAP_BUFFERS | PW_STREAM_FLAG_RT_PROCESS, params, 1);
+    pw_thread_loop_start(state->loop);
+    return 0;
+}
+static int pipewire_play(AudioID *id, AudioTrack track)
+{
+    module_state *state = (module_state *)id;
+    uint32_t write_index;
+    int fill_quantity;
+    // if the stream has been deactivated because of a pipewire_stop call, we activate it
+    if (pw_stream_get_state(state->stream, NULL) == PW_STREAM_STATE_PAUSED)
+    {
+        pw_stream_set_active(state->stream, true);
+    }
+    fill_quantity = spa_ringbuffer_get_write_index(&state->rb, &write_index);
+    // we make sure we don't xrun here at least, aka write somehow beyond the beginning of the buffer, or past its end. This is illogical, but we should still ensure that's not the case
+    spa_assert(fill_quantity >= 0);
+    spa_assert(fill_quantity <= SAMPLE_BUFFER_SIZE);
+    // we write the samples provided in the track we were given, assuming num_samples is the actual length in samples of that memory area, to the ringbuffer, to then be consumed by the on_process callback
+    //  we assume speech dispatcher gives us the correct number of bytes for the format it chose, enough for this chunk. If that's not the case, there's not much we could do besides reading uninitialized memory or an incomplete chunk, unfortunately
+    spa_ringbuffer_write_data(&state->rb, state->sample_buffer, SAMPLE_BUFFER_SIZE, write_index, track.samples, track.num_samples);
+    // now, we update the write index to account for the chunk of samples we just wrote
+    spa_ringbuffer_write_update(&state->rb, write_index + track.num_samples);
```

Comment:
You also need to add a loop with a condition variable: decrement track.num_samples by the amount that could be filled there, and as long as it is not zero, wait for `on_process` to notify. Otherwise we'll be pushing to the frame_buffer only the first part of the audio that can fit, and not wait for pipewire to  play it before pushing the rest.

Reply:
> spd generates enough samples at once to fill that entire buffer?

Of course: speech synthesis produce audio much faster than the audio card plays it.

> how many samples would I typically get at once?

Depends on the syntheses. That's why you shouldn't try to guess and just push to your `frame_buffer` whatever can fit, and then wait for more room to free up before pushing more, etc. until it's all pushed.

> what do you mean by decrementing track.num_samples, I can actually modify that and spd sees it?

No, I rather meant you can make your own copy of it and decrement it.

> I thought that struct is supposed to pass to me as read-only

It's not marked as const, so you can actually modify it. But possibly const was just forgotten, so better make your own copy of the field.

> What should I decrement it by?

The amount you were able to push into `frame_buffer`, i.e. the minimum between its value and  fill_quantity.

Reply:
fixed in the latest commit